idseq_pipeline/commands/curate_accessionid2seq.py

```diff
@@ -0,0 +1,70 @@
+"""The curate_accessionid2seq command."""
+from .base import Base
+import sys
+import time
+import subprocess
+import random
+import pickle
+import shelve
+import argparse
+import gzip
+import re
+import threading
+import traceback
+
+def curate_seq_dict(nt_file, output_db_file):
+    """Curate nt/nr accession id to  file offset to get the sequence"""
+    print "Read the nt/nr file"
+    # Read the nt/nr file
+    taxon_map = shelve.open(output_db_file)
+    with open(nt_file) as ntf:
+        seq_offset = 0
+        seq_len = 0
+        header_len = 0
+        lines = 0
+        accession_id = ""
+        for line in ntf:
+            lines += 1
+            if lines % 100000 == 0:
+                print "%3.1f M lines. " %  (lines/1000000.0)
+            if line[0] == '>': # header line
+                if seq_len > 0 and len(accession_id) > 0:
+                    taxon_map[accession_id] = (seq_offset, header_len, seq_len)
+                seq_offset = seq_offset + header_len + seq_len
+                header_len = len(line)
+                seq_len = 0
+                s = re.match('^>([^ ]*).*', line)
+                if s:
+                    accession_id = s.group(1)
+            else:
+                seq_len += len(line)
+        if seq_len > 0 and len(accession_id) > 0:
+            taxon_map[accession_id] = (seq_offset, header_len, seq_len)
+    taxon_map.close()
+
+class Curate_accessionid2seq(Base):
+    def run(self):
+        from .common import * #TO DO: clean up the imports across this package
+
+        # Make work directory
+        dest_dir = os.path.join(DEST_DIR, 'accession2seq')
+        execute_command("mkdir -p %s" % dest_dir)
+        arguments = self.options
+
+        s3_input_file = arguments.get('--input_s3_path')
+        s3_output_file = arguments.get('--output_s3_path')
+        local_input_file = os.path.join(dest_dir, os.path.basename(s3_input_file))
+        local_output_file = os.path.join(dest_dir, os.path.basename(s3_output_file))
+
+        # Copy the data over from s3
+        execute_command("s3mi cp %s %s" % (s3_input_file, local_input_file))
```

Comment:
Note that s3mi assumes you have at least 24 GB of RAM to spare for buffering.   If your file is bigger than that, and the machine has not enough RAM, s3mi will crash.   If your file is smaller than that, but bigger than the machine RAM, again the machine will crash.   There is an open issue to make s3mi adjust how much RAM it uses based on what's available.

Reply:
Also btw it's hardly worth using `S3MI` if your destination is a single EBS gp2 volume, since you are steady-state bandwidth limited to 160MB/sec by EBS.

Downloading the STAR genome in host filtering takes >15 minutes without S3MI so making it possible to use S3MI for that could help quite a bit.  We'd have to switch to i3.2xlarge (8vCPU, 61 GB RAM) so we can use NVME for the index download.

Reply:
Also it looks like STAR is only using 1 CPU core.

Reply:
It's anecdotally a lot faster than a simple s3cp from copynig NT/NR files.